File: csharp/Magpie.Compilation/Syntax/Pattern Matching/TuplePattern.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Magpie.Compilation
{
    /// <summary>
    /// Pattern for matching a tuple value.
    /// </summary>
    public class TuplePattern : IPattern
    {
        /// <summary>
        /// The patterns for each field in the tuple.
        /// </summary>
        public IList<IPattern> Fields { get; private set; }

        public TuplePattern(IList<IPattern> fields)
        {
            Fields = fields;
        }

        public override string ToString()
        {
            return "(" + Fields.JoinAll(", ") + ")";
        }

        #region ICaseExpr Members

        public Position Position
        {
            get
            {
                return Fields[0].Position;
            }
        }

        TReturn IPattern.Accept<TReturn>(IPatternVisitor<TReturn> visitor)
        {
            return visitor.Visit(this);
        }

        #endregion
    }
}

File: src/com/stuffwithstuff/magpie/ast/pattern/TuplePattern.java
package com.stuffwithstuff.magpie.ast.pattern;

import java.util.List;

import com.stuffwithstuff.magpie.ast.pattern.Pattern;

public class TuplePattern extends Pattern {
  TuplePattern(List<Pattern> fields) {
    mFields = fields;
  }
  
  public List<Pattern> getFields() { return mFields; }
  
  @Override
  public <R, C> R accept(PatternVisitor<R, C> visitor, C context) {
    return visitor.visit(this, context);
  }

  @Override
  public String toString() {
    StringBuilder builder = new StringBuilder();
    for (int i = 0; i < mFields.size(); i++) {
      builder.append(mFields.get(i));
      if (i < mFields.size() - 1) builder.append(", ");
    }
    return builder.toString();
  }
  
  private final List<Pattern> mFields;
}